### PR TITLE
Add a RobotFramework test for c8y-firmware-plugin

### PR DESCRIPTION
**Pull request by rina23q**
## Proposed changes
Create a new test file for c8y-firmware-plugin. Prepared two scenarios:
1. Happy path with file download.
2. Happy path without file download (hitting the file cache).

Note:
* `c8y-firmware-plugin` must be installed after `tedge-agent`, otherwise the installation of `c8y-firmware-plugin` will fail due to the failure to create directories under `/var/tedge` by `c8y-firmware-plugin --init`. That's why the change on the bootstrap script looks a bit strange.

## Types of changes
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#1696 

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: tests/RobotFramework/tests/cumulocity/firmware/firmware_operation_child_device.robot [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    Set Device Context    ${CHILD_SN}
+    Execute Command    sudo rm -f firmware1
+
+Create child device
+    [Documentation]    FIXME: Due to the initialization bug, child device not being registered, we need to restart mapper. See the bug report XXX
+    Set Device Context    ${PARENT_SN}
+    Execute Command    mkdir -p /etc/tedge/operations/c8y/${CHILD_SN}
+    ThinEdgeIO.Transfer To Device    ${CURDIR}/c8y_Firmware    /etc/tedge/operations/c8y/${CHILD_SN}/
+    ThinEdgeIO.Restart Service    tedge-mapper-c8y
+    Cumulocity.Device Should Exist    ${CHILD_SN}
```
Comment (reubenmiller) [inline tests/RobotFramework/tests/cumulocity/firmware/firmware_operation_child_device.robot:102]: I belive this sequence is causing the child device to be registered implicity resulting in the child device name having the `MQTT Device ` prefix in the name.

For example the first test run had a failure when validating the child device name:

```
TST_rev_dense_manuscript != MQTT Device TST_rev_dense_manuscript
```

Comment (rina23q) [reply tests/RobotFramework/tests/cumulocity/firmware/firmware_operation_child_device.robot:102]: Ah horrible, race condition. That means some SmartREST message was published to the child device before `tedge-mapper-c8y` sends 101 to create a child device with appropriate inventory fragments :/

Comment (reubenmiller) [reply tests/RobotFramework/tests/cumulocity/firmware/firmware_operation_child_device.robot:102]: Yep this old chestnut.

Comment (rina23q) [reply tests/RobotFramework/tests/cumulocity/firmware/firmware_operation_child_device.robot]: Yeah, first `114` (declaring supported operation) is arrived, then `101` (creating a child device) is arrived in c8y. Should be opposite order.

After a couple of trials, I figure out that we don't need a restart. Instead, `sleep 3s` after `Set Device Context` helps. From this, I started feeling that the issue is in testing setup environment side. Therefore, I decided not to create an issue yet.

Comment (rina23q) [reply tests/RobotFramework/tests/cumulocity/firmware/firmware_operation_child_device.robot]: > Though today a "sleep 3s" helps, tomorrow it doesn't...makes for a flakey test. I would much prefer to create a dynamic wait instead.

Definitely dynamic wait makes much more sense! Sleep with static seconds is just a workaround...

**Comment by rina23q** [comment]
The test passed with the magic `sleep 3s` :D

File: tests/images/debian-systemd/files/bootstrap.sh [hunk 2]
```diff
@@ -135,6 +135,7 @@ install_via_apt() {
         c8y-configuration-plugin \
         c8y-log-plugin \
         tedge-watchdog
+    apt-get install -y c8y-firmware-plugin
```
Comment (reubenmiller) [inline tests/images/debian-systemd/files/bootstrap.sh:138]: I think we need to raise a ticket to address this dependency issue that we have with the `c8y-firmware-plugin`.

File: tests/RobotFramework/tests/cumulocity/firmware/firmware_operation_child_device.robot [hunk 3]
```diff
@@ -0,0 +1,171 @@
+*** Settings ***
+Resource    ../../../resources/common.resource
+Library    Cumulocity
+Library    ThinEdgeIO
+Library    JSONLibrary
+Library    DebugLibrary
+Library    ../../../.venv/lib/python3.9/site-packages/robot/libraries/String.py
+Library    ../../../.venv/lib/python3.9/site-packages/robot/libraries/OperatingSystem.py
+Library    ../../../.venv/lib/python3.9/site-packages/robot/libraries/DateTime.py
```
[the remark this answers is not in the document]
Comment (rina23q) [reply tests/RobotFramework/tests/cumulocity/firmware/firmware_operation_child_device.robot:9]: Good catch. It needs `String` for `Get Regexp Matches` only.

File: tests/images/debian-systemd/files/bootstrap.sh (hunk 2 above)
Comment (rina23q) [reply tests/images/debian-systemd/files/bootstrap.sh:138]: Created: https://github.com/thin-edge/thin-edge.io/issues/1793

**Review by reubenmiller — APPROVED** [review]
Approved

**Pull request merged**